### ISSUE TITLE
Not receiving rest api web hooks callbacks

### ISSUE THREAD

**Issue by SSMark** [paypal/PayPal-PHP-SDK#242]
Good morning, i'm using sandbox for testing, i developed a simple checkout and a recurring payment, all works fine but i don't receive the webhook callbacks, i made an app inside paypal developer, i made a new sandbox webhook with correct url with all events selected; i made a simple php script that writes on file when the url is visited (works manually and with postman).

I don't know why i'm not receiving the paypal webhook callbacks.

anyone with the same issue?

Regards

**Comment by jaypatel512** [comment]
Hey @SSMark !

Currently, Webhook support is not enabled for subscription APIs, which includes Billing Plan and Agreements. 

And for simple checkout, the webhooks are supported for `paypal` as a payment type which doing the checkout. Can you please confirm if you are using `paypal` as your funding source, and still facing this issue on single checkout flow.

**Comment by avidas** [comment]
@SSMark while work on webhook support for billing agreements is in progress, [IPN](https://github.com/paypal/PayPal-Python-SDK/issues/69#issuecomment-69647946) can a solution in the interim if you want to get notifications.

**Comment by SSMark** [comment]
@jaypatel512  yes i'm using paypal with rest api as funding source, i made a single checkout but no webhook calls.

@avidas thanks for the reply, but i made all the payment system using the rest api, IPN is compatible with rest api? (i think not) i need to delete my payment system and make it again using the classic api?

thanks in advance

**Comment by jaypatel512** [comment]
Hey @SSMark !

I am pretty sure you did that, but webhooks are triggered once execute is done on payment calls. If you have already done that, can you please send me the debug_id of that payment. I could look deeper into the system and see if PayPal is sending the data or not.

What @avidas is saying is, the webhooks are an abstraction layer on top of IPN. So, you could still use the IPN instructions mentioned in [that issue](https://github.com/paypal/PayPal-Python-SDK/issues/69#issuecomment-69647946) to receive webhooks on subscription APIs. You do not need to move away from Rest to Classic. IPN could be used for newer REST subscription API, till the proper support for webhooks for subscription is made available.

**Comment by jaypatel512** [comment]
Hey @SSMark ! Can you please confirm if you are still seeing this issue ? I would be more than happy to help fix that.

**Comment by jaypatel512** [comment]
Closing based on inactivity. Please re-open if still facing the issue.

**Comment by claudio-dalicandro** [comment]
Here we have the same problem, I created a webhook in the sandbox, sending a request for authorization has no effect, of course I expect to arrive after the user has entered the credentials on paypal, but nothing (my logger records the webhook only when I ask that through the webhook test page in the dashboard)

**Comment by liuggio** [comment]
+1 on this...
and is not also possible to match IPN and api rest...

**Comment by jaypatel512** [comment]
Hey @claudio-dalicandro ! 

You will be getting the webhook when the payment/authorization is executed. Once executed, you will be receiving a webhook.

**Comment by jaypatel512** [comment]
@liuggio !

The team is working on getting feature parity between REST webhooks and IPN. However, it would be a gradual increase, as the team is dedicating some time to develop new features too.

**Comment by dopesong** [comment]
@jaypatel512 this shouldn't be closed. It's main of the functionality missing in REST API. Now we have to use both APIs in order to get notifications about payments...

**Comment by snovity** [comment]
@jaypatel512 is there a definitive place to find out which of the following events are working through the API:
```ruby
    'BILLING.PLAN.CREATED',
    'BILLING.PLAN.UPDATED',
    'BILLING.SUBSCRIPTION.CANCELLED',
    'BILLING.SUBSCRIPTION.CREATED',
    'BILLING.SUBSCRIPTION.RE-ACTIVATED',
    'BILLING.SUBSCRIPTION.SUSPENDED',
    'BILLING.SUBSCRIPTION.UPDATED',
    'CUSTOMER.DISPUTE.CREATED',
    'CUSTOMER.DISPUTE.RESOLVED',
    'IDENTITY.AUTHORIZATION-CONSENT.REVOKED',
    'INVOICING.INVOICE.CANCELLED',
    'INVOICING.INVOICE.PAID',
    'INVOICING.INVOICE.REFUNDED',
    'PAYMENT.AUTHORIZATION.CREATED',
    'PAYMENT.AUTHORIZATION.VOIDED',
    'PAYMENT.CAPTURE.COMPLETED',
    'PAYMENT.CAPTURE.DENIED',
    'PAYMENT.CAPTURE.PENDING',
    'PAYMENT.CAPTURE.REFUNDED',
    'PAYMENT.CAPTURE.REVERSED',
    'PAYMENT.PAYOUTSBATCH.DENIED',
    'PAYMENT.PAYOUTSBATCH.PROCESSING',
    'PAYMENT.PAYOUTSBATCH.SUCCESS',
    'PAYMENT.SALE.COMPLETED',
    'PAYMENT.SALE.DENIED',
    'PAYMENT.SALE.PENDING',
    'PAYMENT.SALE.REFUNDED',
    'PAYMENT.SALE.REVERSED',
    'RISK.DISPUTE.CREATED'
```
I'm interested in, for instance, `'PAYMENT.CAPTURE.COMPLETED'`. Right now I do not receive any webhook events for that. Should I try using IPN?

**Comment by jaypatel512** [comment]
Hey @snovity !

Here is the list of all available events: https://developer.paypal.com/docs/integration/direct/rest/webhooks/event-names/

**Comment by usamamashkoor** [comment]
@jaypatel512 it has been two years since this issues has been opened and i am still facing the same issue not receiving any web hook event after the user cancel his subscription from his account or the admin business owner cancels the subscription from his PayPal account.  
i am using webhook event from the available events using the 
https://developer.paypal.com/docs/integration/direct/rest/webhooks/event-names/

event name which i am using is 

BILLING.SUBSCRIPTION.CANCELLED

if this is still not available then please guide me with a alternative.

Thanks

**Comment by xiaoleih41** [comment]
I am very sorry about the situation. The specific event you have mentioned should be available but you should also look into [IPN](https://developer.paypal.com/docs/classic/ipn/integration-guide/IPNIntro/?mark=IPN) as mentioned in the previous thread.

**Comment by usamamashkoor** [comment]
Hi @xiaoleih41 Thanks i can get the PayPal subscription cancel IPN notification and also subscription create notification to through PayPal IPN but i don't see the transaction fees inside the PayPal  subscription created IPN notification how can i get the transaction fee for setup initial fee. 

i know you have created a separated issue for rest api for the transaction fee but i dont know how much time it will take for PayPal developers to show it in response but can you please let me know a alternate solution how can i get the transaction fee for setup initial fee.

**Comment by xiaoleih41** [comment]
@usamamashkoor Most merchant will use PayPal reporting system for accounting reconciliation. In the meantime, please use [PayPal reporting system](https://developer.paypal.com/docs/classic/products/reporting/#online-reports) for the tracking of fees and funds.

**Comment by usamamashkoor** [comment]
@xiaoleih41  but i want to get the transaction fee using api because i want to do the calculations on the amount and we want it dynamically i have seen the initial_payment_txn_id in the IPN response can i get it using this id if yes then kindly guide me a little bit on it if not then please suggest me a way in which i can get the transaction fee dynamically. 
Thanks again.

**Comment by usamamashkoor** [comment]
@xiaoleih41 any help or guidance will be really appreciated.

**Comment by xiaoleih41** [comment]
@usamamashkoor  You can use IPN to get the transaction ID of the the `setup_fee`. Then use it to [get sale detail](https://github.com/paypal/PayPal-PHP-SDK/blob/master/sample/sale/GetSale.php). In the response, you should be able to get the transaction fee.

**Comment by usamamashkoor** [comment]
Thanks @xiaoleih41 for your reply i have also tried another way let me know if this is a also correct because currently i have implemented this technique without paypal IPN.
Here is what i did
after executing the billing agreement using PayPal token i am searching the billing agreement transactions using this code 
```
$params = array('start_date' => date('Y-m-d', strtotime('-2 days')), 'end_date' => date('Y-m-d', strtotime('+2 days')));
$transaction = Agreement::searchTransactions($agreement_id, $params,$api_context);   
 
```
because when the billing agreement is executed successfully their will always be only two transactions one them will be one transaction will have the status created and the second one will have the status completed so the transaction which have status completed is initial setup payment so i have gotten the transaction fee of that transaction which have the status completed because again their can only be one transaction which will have status completed that will always will be initial setup fee because i have set the start date to after 1 month.
let me know if this is the correct way of doing this. 

Thanks in advance.

**Comment by usamamashkoor** [comment]
Thanks @xiaoleih41 for the confirmation.
